app/models/health/qualifying_activity.rb

```diff
@@ -617,13 +617,74 @@ def occurred_within_three_months_of_enrollment?
       date_of_activity.present? && patient.first_n_contributed_days_of_enrollment(90).include?(date_of_activity)
     end
 
+    # at the time of this call does the patient have
+    # a valid care plan covering the date_of_activity
     def patient_has_valid_care_plan?
       return false if patient.care_plan_renewal_date.blank?
       return false unless date_of_activity.present?
 
       date_of_activity >= patient.care_plan_provider_signed_date && date_of_activity < patient.care_plan_renewal_date
     end
 
+
+    # For the date of ac date_of_activity did the patient have
+    # a care plan that MassHealth would consider
+    # valid covering the date_of_activity. This is much
+    # slower and more complex than patient_has_valid_care_plan?
+    # which cares only about the patients status now
+    def patient_had_valid_care_plan?
+      # These have changed over time
+      # FIXME?: do we want to back fill rules in effect
+      # at date_of_activity
+      engagement_period_in_days = ::Health::PatientReferral::ENGAGEMENT_IN_DAYS
+      allowed_gap_in_days = ::Health::PatientReferral::REENROLLMENT_REQUIRED_AFTER_DAYS
+
+      # We are going to need to look at most referrals for this patient
+      patient_referrals = patient.patient_referrals.sort_by(&:enrollment_start_date)
+
+      # Are there any referrals that were active at the time of this activity?
+      # FIXME? Other code assumes that enrollments are non-overlapping
+      # doing so here could make this code faster
+      contributing_referrals = patient_referrals.select do |r|
+        r.active_on?(date_of_activity)
+      end.to_set
+
+      # 0 active referrals means a valid care plan is irreverent/impossible
+      return nil if contributing_referrals.none?
+
+      # Search backward in time and collect any referrals
+      # where the gaps between the its disenrollment_date
+      # and any of our existing contributions is <= allowed_gap_in_days
+      # This is O(n^2) but N is expected to stay small.
+      # FIXME? Other code assumes that enrollments are non-overlapping
+      # doing so here could make this code faster
+      patient_referrals.reverse_each do |r|
+        next if r.in?(contributing_referrals)
+        close_enough = contributing_referrals.any? do |r2|
+          (r2.enrollment_start_date - r.disenrollment_date).to_i <= allowed_gap_in_days
+        end
+        contributing_referrals << r if close_enough
+      end
+
+      # Just in case
+      contributing_referrals = contributing_referrals.to_a.sort_by(&:enrollment_start_date)
+
+      # we need care plan signed within the first 150 accumulated
+      # days of enrollment in the contributing_referrals series
+      enrolled_dates = Set.new
+      contributing_referrals.each do |r|
+        enrolled_dates += r.enrolled_days_to_date
```

Comment:
Does this need to pass in a date? Shouldn't it be looking to see how many days occurred before `date_of_activity`.  Line 649, I think will include referrals that are ongoing on `date_of_activity`, which potentially have end-dates past `date_of_activity`

Reply:
Right, so we are OK to find care plans that occur **after the data of activity** but within the first 150 **accumulated** days of the initial enrollment.

Turns out according to @awisema its 150 accumulated days with an active referral so we have to count the days rather than do date subtraction.

Reply:
Yes, accumulated days.  The care plan could be open after, but the days must have occurred before the QA.

Reply:
If it wasn't clear, I think `r.enrolled_days_to_date` should take a date `r.enrolled_days_to_date(date)` and default to the current logic if none provided, but only count up until the date if one is provided.  OR, you could just exclude any dates `r.enrolled_days_to_date` returns that are after the `date_of_activity`.  The issue is that it may count days after the `date_of_activity` on line 673 even though it throws them out on 675.